Stashing in persistent behavior, #25717

patriknw opened the pull request:
Pushed first wip, but need to think more about this. The order is completely messed up, see problem in the test.

Refs #25717

patriknw committed:
Stashing in persistent behavior, #25717

* Effect.stash, Effect.thenUnstashAll
* unstash one at a time, and start with external stash if UnstashAll is in progress
* handle restarts
* handle PoisonPill
* fix PerformanceSpec, because now the internal stash is cleared if exception is thrown
* work with PoisonPill

patriknw commented:
Pushed a working solution for the ordering problem. Some polish remaining but a good point for a first review.

patriknw commented on akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorStashSpec.scala:
```diff
@@ -0,0 +1,443 @@
+      value3 should ===(8100)
+    }
+
+    // FIXME what is desired behavior?
```
We need to decide if the stash buffers should be cleared or not for restarts.

Note that there are two causes of restarts:
* exception thrown when processing command, handled by outer supervision
* persist failures, handled by `onPersistFailure` backoff supervision

chbatey replied:
This is tricky and I can't decide which way I prefer.

What are the use cases we see stashing being used?
* External resource being down?
* FSM and stashing events until we reach a later state

For `onPersistFailure` I'd likely want to stash to be preserved in both of those cases as we'd restore our state / stash to the same as before the database issue.

For a user exception it I lean the other way as that typically implies we've gotten into a bad state some how. Do I want to keep stashed events that got me there?

When are side effects run like replying to the user? If that is an Ack then it would be surprising if the eventual command were then lost, but the program could crash / drop in the mean time so there's still no guarantee. 

Another option for the user is to model a buffer directly into the state which I think covers the general FSM case.

chbatey replied:
Maybe @renatocaval has some thoughts?

hepin1989 replied:
Can that be overrideable with strategy?

patriknw replied:
> For onPersistFailure I'd likely want to stash to be preserved in both of those cases as we'd restore our state / stash to the same as before the database issue.

I think I agree.

> For a user exception it I lean the other way as that typically implies we've gotten into a bad state some how.

I think I agree. For complete safe stashing the buffer of stashed commands must be persisted instead of using the in-memory buffer, i.e. implemented in application domain.

renatocaval replied:
@chbatey I don't know if I understand you in these two sentences:

> For a user exception it I lean the other way as that typically implies we've gotten into a bad state some how. Do I want to keep stashed events that got me there?

Why do you talk about stashed events? We stash the commands, no? As such, it's ok to preserve the stash after some user exception. The command that would have caused the exception won't be re-processed, but others that were waiting on the queue will have a chance to hit the command handlers. That is fine because we should not make any assumption about any relation between commands. 

If I have `CmdA` and `CmdB` in my mailbox/stash and `CmdA` causes an exception, then we let it crash and we deliver `CmdB` next. It may succeed or not, disregarding what happened with `CmdA`. 

> When are side effects run like replying to the user? If that is an Ack then it would be surprising if the eventual command were then lost, but the program could crash / drop in the mean time so there's still no guarantee.

That, and the previous sentence, makes me think that we want to keep a command around and re-deliver it in case of failures, until it succeeds. 

We could indeed keep it and retry if we can detect that that's only a DB error and not an user exception. That makes sense to me. However, we must be careful with how we deal with applying the events. 

As far as I remember, we do apply the events before we try to persist them. This is useful to check that we are not persisting anything that we can't apply later on. It's an extra guarantee. However, if the persist fails and we try to re-deliver the command we will emit and apply the events once again. 

If the model is immutable, we have protection against it. We can swap old and new state only when the persisting is confirmed, but the problem is that we can't know if the state is immutable.  We may have update it in place and the re-delivery of a failed command will screw up everything. 

Or do we really intent to let it crash and rehydrate the actor while preserving the stash? If that's the case, the state will be discarted and reconstructed on replay. And we will be safe.

patriknw commented on akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorStashSpec.scala:
```diff
@@ -0,0 +1,443 @@
+      ackProbe.expectMessage(Ack("act"))
+      ackProbe.expectMessage(Ack("inc-2"))
+      // inc-3 failed
+      // inc-4, inc-5 still in external stash buffer but UnstashAll in progress was aborted -- not desired
```
I think we could rather easily preserve the complete external stash state for persist failures, if that is what we want

patriknw commented:
This is ready for final review.

This should be merged after https://github.com/akka/akka/pull/26036, and I will deal with the merge conflicts then.

patriknw commented on akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorStashSpec.scala:
```diff
@@ -0,0 +1,487 @@
+    }
+  }
+
+  // FIXME test combination with PoisonPill
```
yeah, I will add that test

I think it should drain any ongoing UnstashAll, but not wait for the external stash to become empty. The reason is that it's known that UnstashAll will complete but it's know that there will be an UnstashAll any time soon

chbatey replied:
:+1:

chbatey commented on akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/PersistentEntity.scala:
```diff
@@ -24,17 +24,17 @@ import akka.persistence.typed.javadsl.PersistentBehavior
   }
 
-  def this(entityTypeKey: EntityTypeKey[Command], entityId: String, backoffSupervisorStrategy: BackoffSupervisorStrategy) = {
+  def this(entityTypeKey: EntityTypeKey[Command], entityId: String, onPersistFailure: BackoffSupervisorStrategy) = {
```
onJournalFailure is maybe a better name. It is also triggered for a replay failing.

patriknw replied:
Not sure the user is aware of the concept of journal. The primary API is Effect.persist and even if we also use it for recovery I think it's easiest to relate to persist.

chbatey commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/SideEffect.scala:
```diff
@@ -42,6 +42,10 @@ final private[akka] class NoReplyEffectImpl[State]
 
+/** INTERNAL API */
+@InternalApi
+private[akka] case object UnstashAll extends SideEffect[Nothing]
```
So Stash is a primary effect and unstash is a side effect? Makes sense.

chbatey commented on akka-docs/src/main/paradox/typed/persistence.md:
````diff
@@ -414,3 +413,30 @@ Journals can reject events. The difference from a failure is that the journal mu
+
+```
+akka.persistence.typed.stash-capacity = 10000
+```
````
Doc ordering of these two types of stashes? Is it that explicit stash is emptied before any of the internal stash is delivered?

chbatey replied:
Doc: What happens if stopped when we have things in the user stash?

chbatey replied:
Doc: If I stash after an unstash all but before they have all been unstashed do i get that stashed command back? Or is that after the next unstash all

chbatey replied:
Doc: What happens for restarts

chbatey commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashManagement.scala:
```diff
@@ -22,16 +22,34 @@ private[akka] trait EventsourcedStashManagement[C, E, S] {
+  protected def isInternalStashEmpty: Boolean = internalStashBuffer.isEmpty
 
-    try stashBuffer.stash(msg) catch {
+  private def externalStashBuffer: StashBuffer[InternalProtocol] = stashState.externalStashBuffer
```
userStash?

chbatey commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashState.scala:
```diff
@@ -0,0 +1,41 @@
+    unstashAllExternalInProgress -= 1
+
+  def startUnstashAllExternal(): Unit =
+    unstashAllExternalInProgress = _externalStashBuffer.size
```
ah so unstashall unstashes the nr at the time it was received :+1: I think that is what I would have guessed

chbatey commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/PersistentBehaviorImpl.scala:
```diff
@@ -76,21 +79,23 @@ private[akka] final case class PersistentBehaviorImpl[Command, Event, State](
             if (signal == PostStop) {
               eventsourcedSetup.cancelRecoveryTimer()
-              clearStashBuffer()
+              stashState.clearStashBuffers()
```
So we went with clearing both stashes and users have to put it in their state if they want to preserve a buffer in their domain?

chbatey replied:
Ah but this is just PostStop, not PreRestart

chbatey commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashState.scala:
```diff
@@ -0,0 +1,41 @@
+  def isUnstashAllExternalInProgress: Boolean =
+    unstashAllExternalInProgress > 0
+
+  def incrementUnstashAllExternalProgress(): Unit =
```
What is this used for? If a user issues a new one half way through would it just work with setting the unstashAllExternalInProgress as it'll get the updated size of the stash?

chbatey commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/Effect.scala:
```diff
@@ -52,6 +52,31 @@ import akka.persistence.typed.ExpectingReply
+   * Stash the current command. Can be unstashed later with `Effect.thenUnstashAll`
+   * or `EffectFactories.unstashAll`.
+   *
+   * * Note that the stashed commands are kept in an in-memory buffer, so in case of a crash they will not be
```
double *

chbatey commented on akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/StashingExample.java:
```diff
@@ -0,0 +1,159 @@
+
+package jdocs.akka.persistence.typed;
+
+import akka.actor.typed.Behavior;
```
Good example

chbatey commented on akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/StashingExample.java:
```diff
@@ -0,0 +1,159 @@
+      }
+    }
+
+    private Effect<Event, State> onEndTask(State state, EndTask command) {
```
I think breaking up the docs to show this piece by piece to highlight the stashing would be more approachable. Not for this PR

chbatey commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashManagement.scala:
```diff
@@ -22,16 +22,34 @@ private[akka] trait EventsourcedStashManagement[C, E, S] {
+  private def stashState: StashState = setup.stashState
 
-  protected def isStashEmpty: Boolean = stashBuffer.isEmpty
+  private def internalStashBuffer: StashBuffer[InternalProtocol] = stashState.internalStashBuffer
```
Maybe inline as we have the StashState class?

chbatey reviewed (commented):
Looking good

patriknw commented:
@chbatey Incorporated feedback

chbatey commented:
Looks like this failure was a race condition in the old impl: 

https://jenkins.akka.io:8498/job/akka-nightly/6111/consoleFull

```
[WARN] [12/05/2018 04:10:25.864] [NullEmptyStateSpec-akka.actor.default-dispatcher-7] [akka.persistence.Persistence(akka://NullEmptyStateSpec)] No default snapshot store configured! To configure a default snapshot-store plugin set the `akka.persistence.snapshot-store.plugin` key. For details see 'reference.conf'
[ERROR] [12/05/2018 04:10:25.877] [NullEmptyStateSpec-akka.actor.default-dispatcher-4] [akka://NullEmptyStateSpec/user/$b] OptionVal.None.get
akka.actor.ActorInitializationException: akka://NullEmptyStateSpec/user/$b: exception during creation
	at akka.actor.ActorInitializationException$.apply(Actor.scala:193)
	at akka.actor.ActorCell.create(ActorCell.scala:669)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:523)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:545)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:283)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.util.NoSuchElementException: OptionVal.None.get
	at akka.util.OptionVal$.get$extension(OptionVal.scala:70)
	at akka.persistence.typed.internal.EventsourcedStashReferenceManagement.stashBuffer(EventsourcedStashReferenceManagement.scala:27)
	at akka.persistence.typed.internal.EventsourcedStashReferenceManagement.stashBuffer$(EventsourcedStashReferenceManagement.scala:21)
	at akka.persistence.typed.internal.PersistentBehaviorImpl.stashBuffer(PersistentBehaviorImpl.scala:35)
	at akka.persistence.typed.internal.PersistentBehaviorImpl.$anonfun$apply$1(PersistentBehaviorImpl.scala:57)
	at akka.persistence.typed.internal.PersistentBehaviorImpl$$Lambda$525/62943530.apply(Unknown Source)
```

chbatey reviewed (approved):
LGTM

patriknw commented:
I think this should be ready for final review (one more approval)

helena commented on akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala:
```diff
@@ -37,7 +37,7 @@ object ClusterShardingPersistenceSpec {
     """
       akka.loglevel = DEBUG
-      #akka.persistence.typed.log-stashing = on
+      akka.persistence.typed.log-stashing = on
```
Why was it commented and now uncommented?

patriknw replied:
leftover from debugging, changed back in 9596cd6

helena commented on akka-docs/src/main/paradox/typed/persistence.md:
```diff
@@ -414,3 +413,44 @@ Journals can reject events. The difference from a failure is that the journal mu
+processed. They will be discarded if the actor (entity) is passivated or rebalanced by Cluster Sharding.
+They will also be discarded if the actor is restarted (or stopped) due to that an exception was thrown
+from processing a command or side effect after persisting. The stash buffer is preserved for persist
+failures if an `onPersistFailure` backoff supervisor strategy is defined.
```
Unclear sentence at least to me
`The stash buffer is preserved for persist failures if an `onPersistFailure` backoff supervisor strategy is defined.`

patriknw replied:
tried to make it more clear in 9596cd6

helena commented on akka-docs/src/main/paradox/typed/persistence.md:
```diff
@@ -414,3 +413,44 @@ Journals can reject events. The difference from a failure is that the journal mu
+Note that the stashed commands are kept in an in-memory buffer, so in case of a crash they will not be
+processed. They will be discarded if the actor (entity) is passivated or rebalanced by Cluster Sharding.
+They will also be discarded if the actor is restarted (or stopped) due to that an exception was thrown
+from processing a command or side effect after persisting. The stash buffer is preserved for persist
```
So:
Commands will be discarded if
* the system crashes
* the stash buffer reaches capacity
* the actor (entity) is passivated or rebalanced by Cluster Sharding
* the actor is restarted (or stopped) because an exception was thrown during processing after persisting

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/SideEffect.scala:
```diff
@@ -58,5 +62,10 @@ object SideEffect {
+  /**
+   * Unstash the commands that were stashed with `javadsl.EffectFactories.stash` or `scaladsl.Effect.stash`
+   */
+  def unstashAll[State](): SideEffect[State] = UnstashAll.asInstanceOf[SideEffect[State]]
```
I just think these functions having `()` are misleading as `def func[A]():SideEffect[A]` because no parameters are accepted. I realize it's done everywhere.

patriknw replied:
Using `()` to indicate that when running the effects they have side effects, in this case the messages in the stash buffer are processed

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/JournalInteractions.scala:
```diff
@@ -25,8 +29,8 @@ private[akka] trait JournalInteractions[C, E, S] {
   protected def internalPersist(
-    state: Running.EventsourcedState[S],
-    event: EventOrTagged): Running.EventsourcedState[S] = {
+    state: Running.RunningState[S],
```
👍  much better

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala:
```diff
@@ -293,34 +306,42 @@ private[akka] object Running {
-    case _ ⇒
-      throw new IllegalArgumentException(s"Not supported side effect detected [${effect.getClass.getName}]!")
+      case _ ⇒
+        throw new IllegalArgumentException(s"Not supported side effect detected [${effect.getClass.getName}]!")
```
The exclamation point isn't necessary

patriknw replied:
fixed

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashManagement.scala:
```diff
@@ -45,34 +62,75 @@ private[akka] trait StashManagement[C, E, S] {
-
-      stashBuffer.unstash(setup.context, behavior.asInstanceOf[Behavior[InternalProtocol]], 1, ConstantFun.scalaIdentityFunction)
+  /**
+   * `tryUnstashOne` is called at the end of processing each command (or when persist is completed)
```
```suggestion
   * `tryUnstashOne` is called at the end of processing each command or when persist is completed.
```

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashManagement.scala:
```diff
@@ -45,34 +62,75 @@ private[akka] trait StashManagement[C, E, S] {
+        if (buffer eq stashState.internalStashBuffer) "internal" else "user", buffer.head)
+
+      if (stashState.isUnstashAllInProgress)
+        stashState.decrementUnstashAllProgress()
```
Is this always a pre-check to `stashState.decrementUnstashAllProgress` or only here? I mean is it something that should not be allowed unless `isUnstashAllInProgress` and it should be a guard vs only separate?

patriknw replied:
Looks like it's better inside `decrementUnstashAllProgress`, fixed in 9596cd6

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashManagement.scala:
```diff
@@ -45,34 +62,75 @@ private[akka] trait StashManagement[C, E, S] {
+  }
+
+  /**
+   * Subsequent `tryUnstashOne` will first drain the user stash buffer, before using the
```
```suggestion
   * Subsequent `tryUnstashOne` will drain the user stash buffer before using the
```

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashManagement.scala:
```diff
@@ -45,34 +62,75 @@ private[akka] trait StashManagement[C, E, S] {
+  /**
+   * Subsequent `tryUnstashOne` will first drain the user stash buffer, before using the
+   * internal stash buffer. It will unstash as many commands as are in the buffer when
+   * `unstashAll` was called, i.e. if subsequent commands stash more those will
```
```suggestion
   * `unstashAll` was called, i.e. if subsequent commands stash more, those will
```

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashManagement.scala:
```diff
@@ -45,34 +62,75 @@ private[akka] trait StashManagement[C, E, S] {
+  def startUnstashAll(): Unit =
+    unstashAllInProgress = _userStashBuffer.size
+
 }
```
This class has so many nested if functions it indicates the design of it internally should be cleaner. Maybe leveraging anonymous functions.

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/Effect.scala:
```diff
@@ -51,14 +52,43 @@ import scala.collection.JavaConverters._
+   * @see [[Effect.thenUnstashAll]]
+   */
+  def unstashAll(): Effect[Event, State] =
+    none().andThen(SideEffect.unstashAll())
```
Are all these no-arg functions with return types using `()` for Java or is it a style choice?

patriknw replied:
Yes, in java it will always be used with `()`

helena commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/Effect.scala:
```diff
@@ -60,14 +60,42 @@ object Effect {
+   * Side effects can be chained with `andThen`
+   */
+  def stash[Event, State](): ReplyEffect[Event, State] =
+    Stash.asInstanceOf[Effect[Event, State]].thenNoReply()
```
Would be nice if we didn't have so much reflection to use the Effect protocol.

helena reviewed (approved):
LGTM just a few questions and comments

patriknw commented on akka-docs/src/main/paradox/typed/persistence.md:
```diff
@@ -414,3 +413,44 @@ Journals can reject events. The difference from a failure is that the journal mu
+:  @@snip [StashingExample.java](/akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/StashingExample.java) { #stashing }
+
+You should be careful to not send more messages to a persistent actor than it can keep up with, otherwise the stash
+buffer will fill up and when reaching its maximum capacity the commands will be dropped. The capacity can be configured with:
```
Feel free to create a ticket for discussion, but that is a general thing that must be thought of when using Actors. Mailboxes may be be unbounded but in practice some flow control is needed to avoid OOME, especially for persistent actors that are typically rather slow.

patriknw committed:
support nested restart and backoff supervision strategies

* problem found by new EventSourcedBehaviorStashSpec
* added tests of this in SupervisionSpec
* the scheduled ScheduledRestart and ResetRestartCount need an
  owner field so that they can be passed on if not scheduled by
  the strategy instance itself
